### PR TITLE
Fix ELF core dump for hybrid and purecap kernels.

### PR DESCRIPTION
**Pull request by qwattash**
Replace __USER_CAP in vn_rdwr with a capability derived from userspace_cap,
this properly generates a capability instead of deriving from a NULL DDC
in the case of a purecap binary on an hybrid kernel.
In the purecap kernel case, the ELF coredump generates the capability.
Add assertion in IOVEC_INIT_C to make sure that the capability given is valid.

### CODE REVIEW

**Comment by jrtc27** [comment]
compress_chunk is still using __USER_CAP so will re-derive even in a purecap kernel. Do we instead want to make core_output, compress_chunk and core_write always take a capability even for a hybrid kernel, given it's a userspace pointer?

File: sys/kern/vfs_vnops.c [hunk 1]
```diff
@@ -549,12 +549,18 @@ vn_rdwr(enum uio_rw rw, struct vnode *vp, void *base, int len, off_t offset,
 		return (EINVAL);
 	auio.uio_iov = &aiov;
 	auio.uio_iovcnt = 1;
-#if __has_feature(capabilities) && !defined(__CHERI_PURE_CAPABILITY__)
+#ifdef __CHERI_PURE_CAPABILITY__
+	IOVEC_INIT_C(&aiov, base, len);
+#else
+#if __has_feature(capabilities)
```
Comment (brooksdavis) [inline sys/kern/vfs_vnops.c:555]: `#elif`

File: sys/sys/_iovec.h [hunk 2]
```diff
@@ -53,6 +53,9 @@ struct iovec {
 	(iovp)->iov_len = (len);					\
 } while(0)
 #define IOVEC_INIT_C(iovp, base, len)	do {				\
```
Comment (brooksdavis) [inline sys/sys/_iovec.h:55]: This should be a separate commit. `(base)` should be assigned to a temporary in case it's an expression with side effects. Is the cast needed now that we have polymorphic builtins?

Comment (jrtc27) [reply sys/sys/_iovec.h:55]: It's also bogus; some (hopefully most?) users of IOVEC_INIT_C pass it a user capability directly, so I can now panic the kernel by calling `getrandom(NULL, 0, 0)`.

**Comment by qwattash** [comment]
> compress_chunk is still using __USER_CAP so will re-derive even in a purecap kernel. Do we instead want to make core_output, compress_chunk and core_write always take a capability even for a hybrid kernel, given it's a userspace pointer?

Right, I missed the one in `compress_chunk`. I previously attempted to do the same thing for hybrid and purecap kernels but felt weird to discard the capability when calling `vn_rdwr_inchunks` from the hybrid kernel and re-derive it. I did not want to mess with the `vn_rdwr` interface. Since there are multiple instances of the same `__USER_CAP` probably passing capabilities around in imgact_elf is cleaner.

File: sys/sys/_iovec.h (hunk 2 above)
Comment (qwattash) [reply sys/sys/_iovec.h:56]: Good point, I'll take it out of the patch.

File: sys/kern/imgact_elf.c [hunk 3]
```diff
@@ -1803,10 +1803,11 @@ core_write(struct coredump_params *p, const void *base, size_t len,
 }
 
 static int
-core_output(void *base, size_t len, off_t offset, struct coredump_params *p,
-    void *tmpbuf)
+core_output(void * __capability base, size_t len, off_t offset,
+    struct coredump_params *p, void *tmpbuf)
 {
 	int error;
+	void *base_vaddr = (void *)(uintptr_t)(uintcap_t)base;
```
Comment (jrtc27) [inline sys/kern/imgact_elf.c:1810]: It's not though for a purecap kernel. Change the name of the function argument to `base_cap` or similar and call this one `base`? Will also reduce the diff.

File: sys/kern/imgact_elf.c [hunk 4]
```diff
@@ -1947,11 +1950,18 @@ __elfN(coredump)(struct thread *td, struct vnode *vp, off_t limit, int flags)
 		Elf_Phdr *php;
 		off_t offset;
 		int i;
+		char * __capability section_addr;
```
Comment (jrtc27) [inline sys/kern/imgact_elf.c:1953]: Not an addr.

File: sys/kern/imgact_elf.c (hunk 3 above)
Comment (qwattash) [reply sys/kern/imgact_elf.c:1810]: Will rename.

File: sys/kern/vfs_vnops.c [hunk 5]
```diff
@@ -549,12 +549,18 @@ vn_rdwr(enum uio_rw rw, struct vnode *vp, void *base, int len, off_t offset,
...
-		IOVEC_INIT_C(&aiov, __USER_CAP(base, len), len);
+		IOVEC_INIT_C(&aiov,
+		    cheri_capability_build_user_data(CHERI_CAP_USER_DATA_PERMS,
+			(vaddr_t)base, len, 0), len);
 	else
-#endif
 		IOVEC_INIT(&aiov, base, len);
+#else
+	IOVEC_INIT(&aiov, base, len);
+#endif
```
Comment (jrtc27) [inline sys/kern/vfs_vnops.c:563]:
```C
#ifdef __CHERI_PURE_CAPABILITY__
	IOVEC_INIT_C(&aiov, base, len);
#else
#if __has_feature(capabilities)
	if (segflg == UIO_USERSPACE)
		IOVEC_INIT_C(&aiov,
		    cheri_capability_build_user_data(CHERI_CAP_USER_DATA_PERMS,
			(vaddr_t)base, len, 0), len);
	else
#endif
		IOVEC_INIT(&aiov, base, len);
#endif
```
is a bit nicer, though why is IOVEC_INIT not the same as IOVEC_INIT_C for a purecap kernel? If it were you could instead have the even nicer:
```C
#if __has_feature(capabilities) && !defined(__CHERI_PURE_CAPABILITY__)
	if (segflg == UIO_USERSPACE)
		IOVEC_INIT_C(&aiov,
		    cheri_capability_build_user_data(CHERI_CAP_USER_DATA_PERMS,
			(vaddr_t)base, len, 0), len);
	else
#endif
		IOVEC_INIT(&aiov, base, len);
```

Comment (qwattash) [reply sys/kern/vfs_vnops.c:563]: Yeah, I agree that it is ugly, in general the IOVEC initializers are almost equivalent for a purecap kernel. The issue in this case is that `IOVEC_INIT` has an assertion on the address that checks that it is not a user address as in general we always want user addresses to be handled as capabilities at this stage. The assertion remains for the purecap kernel `IOVEC_INIT` but in this specific instance we are providing a capability to userland there, so it would fail the assertion.
This is the only instance where we get this issue with iovec.

Comment (jrtc27) [reply sys/kern/vfs_vnops.c:563]: That assertion only makes sense for a hybrid kernel.

Comment (qwattash) [reply sys/kern/vfs_vnops.c:563]: `PTR2CAP` is used in many places where we are not supposed to get user capabilities in the purecap kernel as well though, that is why I did not ifdef it to a no-op. However I can probably completely alias `IOVEC_INIT` and `IOVEC_INIT_C` for a purecap kernel though, as the distinction is relevant only in hybrid.
